src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
     }
 
     /**
+     * Offers direct access to set the configurable list of branch sources <strong>while</strong> preserving
+     * branch source id associations for sources that are otherwise unmodified
+     * @param sources
+     * @throws IOException
+     */
+    public void setSourcesList(List<BranchSource> sources) throws IOException {
+        if (this.sources.isEmpty() || sources.isEmpty()) {
+            // easy
+            this.sources.replaceBy(sources);
+            return;
+        }
+        Set<String> oldIds = sourceIds(this.sources);
+        Set<String> newIds = sourceIds(sources);
+        if (oldIds.containsAll(newIds) || newIds.containsAll(oldIds)) {
+            // either adding, removing, or updating without an id change
+            this.sources.replaceBy(sources);
+            return;
+        }
+        // Now we need to check if any of the new entries are effectively the same as an old entry that is being removed
+        // we will store the ID changes in a map and process all the affected branches to update their sourceIds
+        Map<String,String> changedIds = new HashMap<>();
+        Set<String> additions = new HashSet<>(newIds);
+        additions.removeAll(oldIds);
+        Set<String> removals = new HashSet<>(oldIds);
+        removals.removeAll(newIds);
+
+        // we will be manipulating the sources, so make sure we own the backing list impl
+        sources = new ArrayList<>(sources);
+        for (ListIterator<BranchSource> i = sources.listIterator(); i.hasNext(); ) {
+            BranchSource addition = i.next();
```

Comment:
I had in my mind to replace parts of the list as I iterated... for the life of me I cannot fathom why. Will change back (and remove the ArrayList 3 lines above)